R/dalys.R

```diff
@@ -0,0 +1,169 @@
+  dalys_src$duration <-  dalys_src$average_duration * duration_weighting(dalys_src$period)
+  
+  
+  i <- dalys_src$period == "year"
+  dalys_src$duration[i] <- dalys_src$average_duration[i]
+  i <- dalys_src$period == "month"
+  dalys_src$duration[i] <- dalys_src$average_duration[i] / 12
+  i <- dalys_src$period == "day"
+  dalys_src$duration[i] <- dalys_src$average_duration[i] / 365
+  
+  ## adjusted weight = proportion * diability_weight; this saves calculation later
+  dalys_src$adjusted_weight <- dalys_src$proportion * dalys_src$disability_weight
+  
+  ## find relevant responsibilities
+  sql <- read_sql(file_name = "dalys_calculation/responsibility.sql")
+  recipe <- DBI::dbGetQuery(con, sql, touchstone_name)
+  
+  dalys_src <- merge(dalys_src, recipe, by = c("disease", "modelling_group"), all.x = TRUE)
+  
+  dalys_src <- dalys_src[order(dalys_src$burden_estimate_set_id), ]
+  cols <- c("disease", "modelling_group", "scenario", "burden_estimate_set_id","burden_outcome_id", "code", "sub_condition", "duration", "adjusted_weight")
+  ## make a temporary table - dalys_parameters. This stores model-specific dalys parameters.
+  DBI::dbWriteTable(con, "dalys_parameters", dalys_src[cols], temporary = TRUE, append = TRUE, row.names = FALSE)
+  dalys_src[cols]
+}
+
+create_dalys_life_table <- function(con, touchstone_name = "201710gavi", year_min = 2000, year_max = 2030) {
+  message("Creating touchstone specific life table - remaining life expectancy by country, year and age")
+  sql <- read_sql(file_name = "dalys_calculation/raw_remaining_life_expectancy.sql")
+  
+  life_ex <- DBI::dbGetQuery(con, sql, touchstone_name)
+  cols <- c("country", "gender", "year", "age", "value")
+  
+  # interpolate years
+  l <- split(life_ex, list(life_ex$country, life_ex$age_from, life_ex$gender))
+  life_ex2 <- do.call(rbind, lapply(l, function(i) interpolate_year(i, year_min, year_max)))
+  
+  # interpolate age
+  # keep age 0
+  life_ex0 <- life_ex2[life_ex2$age == 0,]
+  # after age 0, remaining life expectancy is monotonic
+  life_ex3 <- life_ex2[life_ex2$age != 0,]
+  l <- split(life_ex3, list(life_ex3$country, life_ex3$year, life_ex3$gender))
+  life_ex4 <- do.call(rbind, lapply(l, function(i) interpolate_age(i)))
+  # bind age 0 and other ages
+  dat <- rbind(life_ex0[cols],life_ex4[cols])
+  # output
+  dat$.code <- paste(dat$country, dat$year, dat$age, sep="-")
+  dat$remainning_life_exp <- dat$value
+  dat <- dat[c(".code", "gender", "remainning_life_exp")]
+  dat
+}
+
+interpolate_year <- function(dat, year_min=2000, year_max=2101) {
+  ## expand life table regarding year - from 5year interval to 1year
+  years <- year_min:year_max
+  f <- splinefun(x=dat$year, y=dat$value, method = "natural") #natural spline
+  interp <- data_frame(year = years)
+  interp$value <- f(years)
+  interp$country <- unique(dat$country)
+  interp$gender <- unique(dat$gender)
+  interp$age <- unique(dat$age_from)
+  interp
+}
+
+interpolate_age <- function(dat) {
```

Comment:
This function, as written, assumes that there is only one unique value of country, gender and year, but potentially more than one of age.  So replace `unique` for country/gender/year with `dat$country[[1]]` (etc)

Reply:
Done.